File: Linear Control Course/Chapter4/Lesson1/Chapter4_Lesson1.cpp
#include <iostream>
#include <vector>
#include <complex>

class TransferFunction {
public:
    // G(s) = num(s) / den(s)
    std::vector<double> num;
    std::vector<double> den;

    TransferFunction(const std::vector<double>& n,
                     const std::vector<double>& d)
        : num(n), den(d) {}

    std::complex<double> eval(const std::complex<double>& s) const {
        std::complex<double> N(0.0, 0.0), D(0.0, 0.0);
        for (std::size_t i = 0; i < num.size(); ++i) {
            int p = static_cast<int>(num.size() - 1 - i);
            N += num[i] * std::pow(s, p);
        }
        for (std::size_t i = 0; i < den.size(); ++i) {
            int p = static_cast<int>(den.size() - 1 - i);
            D += den[i] * std::pow(s, p);
        }
        return N / D;
    }
};

int main() {
    double K = 2.0;
    double T = 0.5;

    std::vector<double> num{K};
    std::vector<double> den{T, 1.0};

    TransferFunction G(num, den);

    // Evaluate G(j*omega) at omega = 1 rad/s (for analysis)
    std::complex<double> s(0.0, 1.0);
    std::complex<double> Gjw = G.eval(s);

    std::cout << "G(j*1) = " << Gjw << std::endl;

    // In a robotics context, this class can be used inside a ROS node
    // to approximate continuous dynamics or to design simple controllers.
    return 0;
}

File: Tutorials/Codes/Linear_Control_Course/Chapter4/Lesson1/Chapter4_Lesson1.java
public class TransferFunction {
    // G(s) = num(s) / den(s)
    private double[] num;
    private double[] den;

    public TransferFunction(double[] num, double[] den) {
        this.num = num;
        this.den = den;
    }

    // Evaluate G(s) for complex s = sigma + j*omega
    public Complex eval(Complex s) {
        Complex N = Complex.ZERO;
        Complex D = Complex.ZERO;

        for (int i = 0; i < num.length; i++) {
            int p = num.length - 1 - i;
            N = N.add(s.pow(p).multiply(num[i]));
        }
        for (int i = 0; i < den.length; i++) {
            int p = den.length - 1 - i;
            D = D.add(s.pow(p).multiply(den[i]));
        }
        return N.divide(D);
    }

    public static void main(String[] args) {
        double K = 2.0;
        double T = 0.5;
        double[] num = {K};
        double[] den = {T, 1.0};

        TransferFunction G = new TransferFunction(num, den);
        Complex s = new Complex(0.0, 1.0); // j*1
        Complex Gjw = G.eval(s);
        System.out.println("G(j*1) = " + Gjw);
    }
}
